components/application-connectivity-validator/internal/validationproxy/handler.go

```diff
@@ -0,0 +1,179 @@
+package validationproxy
+
+import (
+	"crypto/x509/pkix"
+	"net/http"
+	"net/http/httputil"
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/kyma-project/kyma/components/application-connectivity-validator/internal/httptools"
+
+	"github.com/kyma-project/kyma/components/application-connectivity-validator/internal/apperrors"
+
+	"github.com/gorilla/mux"
+)
+
+const (
+	CertificateInfoHeader = "X-Forwarded-Client-Cert"
+)
+
+type ProxyHandler interface {
+	ProxyAppConnectorRequests(w http.ResponseWriter, r *http.Request)
+}
+
+type proxyHandler struct {
+	group                  string
+	tenant                 string
+	eventServicePathPrefix string
+	eventServiceHost       string
+	appRegistryPathPrefix  string
+	appRegistryHost        string
+}
+
+func NewProxyHandler(group, tenant, eventServicePathPrefix, eventServiceHost, appRegistryPathPrefix, appRegistryHost string) *proxyHandler {
+	return &proxyHandler{
+		group:                  group,
+		tenant:                 tenant,
+		eventServicePathPrefix: eventServicePathPrefix,
+		eventServiceHost:       eventServiceHost,
+		appRegistryPathPrefix:  appRegistryPathPrefix,
+		appRegistryHost:        appRegistryHost,
+	}
+}
+
+func (ph *proxyHandler) ProxyAppConnectorRequests(w http.ResponseWriter, r *http.Request) {
+	certInfoData := r.Header.Get(CertificateInfoHeader)
+	applicationName := mux.Vars(r)["application"]
+
+	if applicationName == "" {
+		httptools.RespondWithError(w, apperrors.BadRequest("Application name not specified"))
+		return
+	}
+
+	log.Infof("Proxying request for %s application. Path: %s", applicationName, r.URL.Path)
+
+	subjects := extractSubjects(certInfoData)
+	if !hasValidSubject(subjects, applicationName, ph.group, ph.tenant) {
+		httptools.RespondWithError(w, apperrors.Forbidden("No valid subject found"))
+		return
+	}
+
+	host, err := ph.determineHost(r.URL.Path)
+	if err != nil {
+		httptools.RespondWithError(w, err)
+		return
+	}
+
+	reverseProxy := &httputil.ReverseProxy{
```

Comment:
I would suggest to cache ReverseProxy objects here.

Reply:
As discussed, we actually need only two instances of ReverseProxy, therefore there is no need for caching.